README.md
## BeeCloud Python SDK (Open Source)

[![Build Status](https://travis-ci.org/beecloud/beecloud-python.svg)](https://travis-ci.org/beecloud/beecloud-python) ![license](https://img.shields.io/badge/license-MIT-brightgreen.svg) ![version](https://img.shields.io/badge/version-v3.5.0-blue.svg)

## 简介

本项目的官方GitHub地址是 [https://github.com/beecloud/beecloud-python](https://github.com/beecloud/beecloud-python)

SDK支持以下支付渠道: 
 
 * 微信
 * 支付宝
 * 银联
 * PayPal
 * 百度钱包   
 * 京东
 * 易宝、快钱等

并且包含退款、打款、订阅支付和相应的查询功能，适用于python2.7、python3.4、python3.5、python3.6。


## 安装 Python SDK
从pip快速安装  
`pip install beecloud`  
更新  
`pip install beecloud --upgrade`  
或者下载源码进入sdk安装  
`python setup.py install`  


## 依赖
sdk依赖于开源库[requests](http://docs.python-requests.org/en/latest/)，安装sdk时会自动安装，如果意外安装失败，可以手动安装

```
pip install requests
```

demo依赖于开源web框架[Flask](http://flask.pocoo.org/)，需手动安装

```
pip install Flask
```

## 准备
参考[快速开始](https://beecloud.cn/apply/)，完成开发准备工作。

## 流程

下图为整个支付的流程:
![pic](http://7xavqo.com1.z0.glb.clouddn.com/img-beecloud%20sdk.png)


步骤①：**（从网页服务器端）发送订单信息**  
步骤②：**收到BeeCloud返回的渠道支付地址（比如支付宝的收银台）**  
>*特别注意：
微信扫码返回的内容不是和支付宝一样的一个有二维码的页面，而是直接给出了微信的二维码的内容，需要用户自己将微信二维码输出成二维码的图片展示出来*

步骤③：**将支付地址展示给用户进行支付**  
步骤④：**用户支付完成后通过一开始发送的订单信息中的return_url来返回商户页面**
>*特别注意：
微信没有自己的页面，二维码展示在商户自己的页面上，所以没有return url的概念，需要商户自行使用一些方法去完成这个支付完成后的跳转（比如后台轮询查支付结果）*

此时商户的网页需要做相应界面展示的更新（比如告诉用户"支付成功"或"支付失败")。**不允许**使用同步回调的结果来作为最终的支付结果，因为同步回调有极大的可能性出现丢失的情况（即用户支付完没有执行return url，直接关掉了网站等等），最终支付结果应以下面的异步回调为准。

步骤⑤：**（在商户后端服务端）处理异步回调结果（[Webhook](https://beecloud.cn/doc/?index=webhook)）**
 
付款完成之后，根据客户在BeeCloud后台的设置，BeeCloud会向客户服务端发送一个Webhook请求，里面包括了数字签名，订单号，订单金额等一系列信息。客户需要在服务端依据规则要验证**数字签名是否正确，购买的产品与订单金额是否匹配，这两个验证缺一不可**。验证结束后即可开始走支付完成后的逻辑。

## 使用方法
* 具体使用请参考项目中的`demo`代码；
* 关于字符串的说明，对于`python2`如果需要传入的参数包含中文字符，请传入`unicode`，对于网络请求成功的情况，`BCResult`中返回结果的字符串也是`unicode`形式；对于`python3`，不需要考虑这样的细节；
* 以下的示例和`demo`中出现的关键字`u`（`unicode`）是为了兼容处理，在`python3`环境下不需要作这样的处理；
* 关于请求参数，公用字段（`app_id`，`timestamp`，`app_sign`）会自动处理，不要手动设置，其他字段和`REST API`一致（例如`REST API`中支付部分，对于`WX_JSAPI`支付方式，`openid`是必填的，假设请求参数名为`req_params`，那么应该添加这样的设置 `req_params.openid = 'openid_str'`），请参考[官网](https://beecloud.cn/doc/?index=rest-api)说明，境外支付请参考[Github beecloud-rest-api](https://github.com/beecloud/beecloud-rest-api/)，以下不做额外介绍
* 返回结果是`beecloud.entity.BCResult`实例，包含以下公共字段，其他字段因不同API而异（例如`REST API`中支付部分，支付完成后会返回支付表记录唯一标识`id`，假设返回参数名为`result`，可以通过`result.id`获取结果），同理，请参照上一条列出的文档

参数名 | 说明
---- | -----
result_code | 返回码，0为正常
result_msg | 返回信息， OK为正常
err_detail | 具体错误信息

### 1.初始化

#### ①. `BCApp`对应于`BeeCloud`控制台中的应用，初始化：  

```
from beecloud.entity import BCApp
bc_app = BCApp()
bc_app.app_id = 'your app id'
# app secret被用于支付和查询
bc_app.app_secret = 'your app secret'
# master secret被用于打款和退款
bc_app.master_secret = 'your app master secret'
```

如果相关渠道尚未申请完成，想要体验sdk支付流程，可以开启测试模式

```
from beecloud.entity import BCApp
bc_app = BCApp()
# 不要在上线模式中设置该选项，否则会遇到BC验签错误相关提示
app.is_test_mode = True
bc_app.app_id = 'your app id'
bc_app.test_secret = 'your app test secret'
```

#### ②. 对于支付、打款和退款，需要初始化BCPay：

```python
from beecloud.pay import BCPay
bc_pay = BCPay()
bc_pay.register_app(bc_app)
```

#### ③. 对于查询需要初始化BCQuery：

```python
from beecloud.query import BCQuery
bc_query = BCQuery()
bc_query.register_app(bc_app)
```

### 2.支付
可以参考`demo.py`中`app_bill`

#### 原型：
通过`BCPay`的实例，以`pay`方法，结合`BCPayReqParams`参数，发起支付请求；  
对于线下收款，通过`offline_pay`方法发起请求，用户的支付结果通过`BCQuery`的`query_offline_bill_status`方法查询

#### 调用：

```python
req_params = BCPayReqParams()
req_params.channel = 'UN_WEB'
req_params.title = u'支付测试'
# 分为单位
req_params.total_fee = 1
req_params.bill_no = 'billno123'
# 可选参数
req_params.optional = {'lang': 'python', u'中文key': u'中文value'}
# 支付完成后的跳转页面
req_params.return_url = 'https://beecloud.cn/'
result = bc_pay.pay(req_params)
# 如果result.result_code为0表示请求成功
# 然后对相关的返回参数做处理，比如ALI_WEB会返回重定向url
```


### 3.退款
可以参考`demo.py`中`app_refund`；<br/>
退款接口包含预退款功能，当need_approval字段的值为true时，该接口开启预退款功能，预退款仅创建退款记录，并不真正发起退款，需后续调用审核接口

#### 原型：
通过`BCPay`的实例，以`refund`方法，结合`BCRefundReqParams`参数，发起退款请求

#### 调用：

```python
refund_params = BCRefundReqParams()
# 退款channel为选填参数
refund_params.channel = 'WX'
refund_params.refund_no = 'refundno123'
refund_params.bill_no = 'billno123'
# 分为单位
refund_params.refund_fee = 1
# need_approval为True时表示预退款，需要后期调用 预退款批量审核 API
# refund_params.need_approval = True
result = bc_pay.refund(refund_params)
# 如果result.result_code为0表示请求成功
# 对于支付宝退款，需要重定向至result.url
```


### 4.预退款批量审核
可以参考`demo.py`中`app_audit_pre_refunds`

#### 原型：
通过`BCPay`的实例，以`audit_pre_refunds`方法，结合`BCPreRefundAuditParams`参数，发起预退款批量审核

#### 调用：

```python
req_params = BCPreRefundAuditParams()
req_params.channel = 'WX'
req_params.ids = ['pre_refund_id1', 'pre_refund_id2']
req_params.agree = True
result = bc_pay.audit_pre_refunds(req_params)
# result.result_map为批量同意单笔结果集合
# 对于支付宝，需要重定向至result.url
```


### 5.打款
可以参考`demo.py`中`app_transfer`

#### 原型：
打款分**单笔打款**、**比可企业打款**、**批量打款**；  

 * **单笔打款**包含`WX_REDPACK`（微信红包）、`WX_TRANSFER`（微信企业打款）和`ALI_TRANSFER`（支付宝企业打款），通过`BCPay`的实例，以`transfer`方法，结合`BCTransferReqParams`参数发起打款；  

 * **比可企业打款**通过`BCPay`的实例，以`bc_transfer`方法，结合`BCCardTransferParams`参数发起打款；该打款渠道支持的银行列表，可以通过`BCQuery`中的`query_bc_transfer_supported_banks`方法进行查询；
  
 * **批量打款**目前只支持`ALI`（支付宝批量打款），通过`BCPay`的实例，以`batch_transfer`方法，结合`BCBatchTransferParams`参数发起打款；

#### 调用：
***单笔打款***  
以微信红包为例

```python
transfer_params = BCTransferReqParams()
transfer_params.channel = 'WX_REDPACK'
transfer_params.desc = u'打款说明'
# 微信为10位数字
transfer_params.transfer_no = '0123456789'
# 微信红包1.00-200元，此处以分为单位
transfer_params.total_fee = 100
transfer_params.channel_user_id = 'receiver_wechat_open_id'
# 微信红包初始化BCTransferRedPack
redpack = BCTransferRedPack()
redpack.send_name = 'BeeCloud'
redpack.wishing = u'BeeCloud祝福开发者工作顺利'
redpack.act_name = u'BeeCloud开发者测试中'
transfer_params.redpack_info = redpack
result = bc_pay.transfer(transfer_params)
# result.result_code等于0表示打款成功
# 对于支付宝需要重定向到result.url
```
  
***比可企业打款***

```python
transfer_params = BCCardTransferParams()
# 单位为分
transfer_params.total_fee = 1
transfer_params.bill_no = order_num_on_datetime()
# 最长支持16个汉字
transfer_params.title = u'python比可企业打款测试'
# 银行缩写编码
transfer_params.bank_code = 'BOC'
# 银行联行行号
transfer_params.bank_associated_code = '1043050000'
# 银行全名
transfer_params.bank_fullname = u'中国银行'
# DE代表借记卡，CR代表信用卡
transfer_params.card_type = 'DE'
# 帐户类型，P代表私户，C代表公户
transfer_params.account_type = 'C'
# 收款方的银行卡号
transfer_params.account_no = '5300000'
# 收款方的姓名或者单位名
transfer_params.account_name = u'苏州比可网络科技有限公司'
# 银行绑定的手机号，当需要手机收到银行入账信息时，该值必填，前提是该手机在银行有短信通知业务，否则收不到银行信息
transfer_params.mobile = '1850000'
# 附加数据，选填
transfer_params.optional = {'key1': u'选填的value'}

result = bc_pay.bc_transfer(transfer_params)
# result.result_code等于0表示打款请求成功，但是需要在webhook判定最终打款结果
```
  
***批量打款***

```python
transfer_params = BCBatchTransferParams()
# 11-32位数字字母组合
transfer_params.batch_no = 'abcdefg1234567'
transfer_params.account_name = u'苏州比可网络科技有限公司'
# 每一笔打款细项为BCBatchTransferItem实例
item1 = BCBatchTransferItem()
item1.transfer_id = order_num_on_datetime() + 'a'
item1.receiver_account = '1234567'
item1.receiver_name = u'某人1'
# 以分为单位
item1.transfer_fee = 1
item1.transfer_note = u'python支付宝批量打款item1'

item2 = BCBatchTransferItem()
# 继续像item1一样添加相关设置

transfer_params.transfer_data = [item1, item2]

result = bc_pay.batch_transfer(transfer_params)
# result.result_code等于0时需要重定向到result.url
```


### 6.查询支付订单
可以参考`demo.py`中`app_query_bills`

#### 原型：
通过`BCQuery`的实例，以`query_bills`方法，结合`BCQueryReqParams`参数查询

#### 调用：
```python
query_params = BCQueryReqParams()
# 如果查询全部订单channel不设置即可
query_params.channel = 'WX'
# 限制只返回支付成功的订单
query_params.spay_result = True
result = bc_query.query_bills(query_params)
# 如果查询成功result.bills为beecloud.entity.BCBill的实例列表
```


### 7.查询退款订单
可以参考`demo.py`中`app_query_refunds`

#### 原型：
通过`BCQuery`的实例，以`query_refunds`方法，结合`BCQueryReqParams`参数查询

#### 调用：
```python
query_params = BCQueryReqParams()
# 如果查询全部订单channel不设置即可
query_params.channel = 'WX'
result = bc_query.query_refunds(query_params)
# 如果查询成功result.refunds为beecloud.entity.BCRefund的实例列表
```


### 8.查询支付订单数目
可以参考`demo.py`中`app_query_bills_count`

#### 原型：
通过`BCQuery`的实例，以`query_bills_count`方法，结合`BCQueryReqParams`参数查询

#### 调用：
```python
query_params = BCQueryReqParams()
# 如果查询全部订单channel不设置即可
query_params.channel = 'WX'
query_params.spay_result = True
result = bc_query.query_bills_count(query_params)
# 如果查询成功result.count表示满足条件的数目
```


### 9.查询退款订单数目
可以参考`demo.py`中`app_query_refunds_count`

#### 原型：
通过`BCQuery`的实例，以`query_refunds_count`方法，结合`BCQueryReqParams`参数查询

#### 调用：
```python
query_params = BCQueryReqParams()
# 如果查询全部订单channel不设置即可
query_params.channel = 'WX'
result = bc_query.query_refunds_count(query_params)
# 如果查询成功result.count表示满足条件的数目
```


### 10.根据ID查询支付订单
可以参考`demo.py`中`app_bill_by_id`

#### 原型：
通过`BCQuery`的实例，以`query_bill_by_id`方法，结合支付订单唯一标识id查询

#### 调用：
```python
result = bc_query.query_bill_by_id(bill_id)
# 如果查询成功result.pay为beecloud.entity.BCBill的实例
```


### 11.根据ID查询退款订单
可以参考`demo.py`中`app_refund_by_id`

#### 原型：
通过`BCQuery`的实例，以`app_refund_by_id`方法，结合退款订单唯一标识id查询

#### 调用：
```python
result = bc_query.query_refund_by_id(refund_id)
# 如果查询成功result.refund为beecloud.entity.BCRefund的实例
```


### 12.退款状态更新
可以参考`demo.py`中`app_query_refund_status`；  
退款状态更新接口提供查询退款状态以更新退款状态的功能，用于对退款后不发送回调的渠道（WX、YEE、KUAIQIAN、BD）退款后的状态更新

#### 原型：
通过`BCQuery`的实例，以`query_refund_status`方法，结合渠道类型channel和退款单号refund_no查询

#### 调用：
```python
result = bc_query.query_refund_status(channel, refund_no)
# 如果查询成功result.refund_status为查询到的退款状态
```

  
### 13.订阅支付
先查看[订阅系统说明文档](https://github.com/beecloud/beecloud-rest-api/blob/master/subscription/%E8%AE%A2%E9%98%85%E7%B3%BB%E7%BB%9F%E8%AF%B4%E6%98%8E%E6%96%87%E6%A1%A3.md)了解基础概念  

* **查询计划列表**
  
可以参考`demo.py`中`subscription_plans`  
  
**原型**
  
通过`BCQuery`的实例，以`query_plans`方法，结合`BCQueryCriteria`参数查询，结果包含`BCPlan`的列表；如果设置了`BCQueryCriteria count_only`为`True`，那么结果只包含满足条件的个数`total_count`  
  
**调用**
  
```python
result = bc_query.query_plans()
# 如果查询成功，result.plans为查询到的计划列表
```
  
* **订阅支付支持的银行列表**
  
可以参考`demo.py`中`subscription_supported_banks`  
  
**原型**
  
通过`BCQuery`的实例，结合`query_subscription_payment_supported_banks`方法发起请求
  
**调用**
  
```python
result = bc_query.query_subscription_payment_supported_banks()
# 如果查询成功，result.common_banks包含支持的常用银行列表，result.banks包含全部支持的银行列表
```
  
* **发送验证码**
  
可以参考`demo.py`中`sms`  
  
**原型**
  
通过`BCPay`的实例，结合`send_sms_passcode`方法发起请求，参数接受验证码的手机号，返回结果包含`sms_id`，同时发送`sms_code`到用户手机  
  
**调用**
  
```python
result = bc_pay.send_sms_passcode(mobile)
# 如果请求成功，result.sms_id为本次验证码id
```
  
* **发起订阅**
  
可以参考`demo.py`中`subscribe`  
  
**原型**
  
通过`BCPay`的实例，以`subscribe`方法，结合`BCSubscription`参数、`sms_id`和`sms_code`发起，结果包含`BCSubscription`对象，如果该对象`valid`为`True`表明本次订阅已经即时生效，否则你还需要等待webhook推送最终审核结果；  

changelog.txt
BeeCloud Python Changelog
=============================
v3.5.0 发布日期：2017年7月12日
添加用户系统支持，商户可以在比可注册自己的用户后，在控制台查看用户行为分析

v3.4.1 发布日期：2017年6月22日
在bc analysis添加sdk version

v3.4.0 发布日期：2016年11月2日
添加线下支付接口

v3.3.0 发布日期：2016年7月7日
根据rest API升级错误处理
添加订阅支付
添加BC_NATIVE支付渠道和鉴权接口

v3.2.0 发布日期：2016年6月8日
+ 比可企业打款支持的银行列表查询API：query_bc_transfer_supported_banks
+ UN_WAP支付

v3.1.0 发布日期：2016年5月20日
+ BeeCloud网关支付
+ BeeCloud快捷支付

v3.0.0 发布日期：2016年1月15日
本次更新比较大，API不兼容v2版本，具体使用查看readme

v2.0.1 发布日期：2015年8月19日
+ 退款不必指定channel
+ 查询订单不必指定channel
+ 查询退款不必指定channel
+ 去掉api.beecloud.cn后端地址

v2.0.0 发布日期：2015年7月29日
+ 随机选取后台服务器进行数据请求
+ 5个rest api请求函数: 支付， 退款， 订单查询， 退款查询， 退款状态更新； 接口参数统一
+ 微信下单接口的返回值有变动，老版本用户请参考demo进行修改
+ 基于tornado的 python web demo, 包含5大功能

v1.2.3 发布日期：2015年4月17日
+ 微信红包功能
+ 修复一些bug

v1.2.1 发布日期：2015年4月10日
+ 修复一些bug

v1.2.0 发布日期：2015年4月8日
+ 加入微信，支付宝，银联支付渠道
+ 加入pypi支持


demo/demo.py
# -*- coding: utf-8 -*-
"""
    This module shows sdk usage.
    :created by xuanzhui on 2015/12/25.
    :copyright (c) 2015 BeeCloud.
    :license: MIT, see LICENSE for more details.
"""
from flask import Flask, request, redirect, render_template, Markup
import datetime
# 实际项目中建议按需import
from beecloud.pay import BCPay
from beecloud.query import BCQuery
from beecloud.subscribe import BCSubscribe
from beecloud.utils import order_num_on_datetime, local_timestamp_since_epoch, fetch_code, fetch_open_id, \
    send_sms_passcode
from beecloud.entity import BCApp, BCPayReqParams, BCRefundReqParams, BCChannelType, BCInternationalPayParams, \
    BCQueryReqParams, BCPreRefundAuditParams, BCBatchTransferParams, BCBatchTransferItem, BCTransferReqParams, \
    BCTransferRedPack, BCCardTransferParams, BCSubscription, BCQueryCriteria
import json

app = Flask(__name__)

# init
bc_app = BCApp()
# bc_app.is_test_mode = True
bc_app.app_id = 'c5d1cba1-5e3f-4ba0-941d-9b0a371fe719'
bc_app.app_secret = '39a7a518-9ac8-4a9e-87bc-7885f33cf18c'
bc_app.master_secret = 'e14ae2db-608c-4f8b-b863-c8c18953eef2'
# bc_app.test_secret = '4bfdd244-574d-4bf3-b034-0c751ed34fee'

# 以下是jsapi的测试参数
wx_appid = "wx119a2bda81854ae0"
wx_app_secret = "53e3943476118a3dff21fb95848de6d7"
redirect_url = "http://pythondemo.beecloud.cn/bill"
# jsapi测试参数结束

bc_pay = BCPay()
bc_pay.register_app(bc_app)

bc_query = BCQuery()
bc_query.register_app(bc_app)

bc_subscribe = BCSubscribe()
bc_subscribe.register_app(bc_app)


@app.route('/')
def hello_index():
    return app.send_static_file('index.html')


@app.route('/bill', methods=['POST', 'GET'])
def app_bill():
    try:
        channel = request.form['channel']
    except:
        channel = request.args.get('channel')
    if channel.startswith('PAYPAL'):
        return _deal_with_international_pay(channel)
    elif channel == 'WX_JSAPI':
        code = request.args.get('code', '')
        if not code:
            oauth_url = fetch_code(wx_appid, redirect_url + '?channel=' + channel)
            return redirect(oauth_url)
        else:
            open_id = fetch_open_id(wx_appid, wx_app_secret, code)
            return _deal_with_normal_pay(channel, open_id)
    elif channel == 'BC_GATEWAY':
        bank = request.form.get('bank')
        if not bank:
            return render_template('choose_bank.html', banks=bc_query.query_bc_gateway_supported_banks())
        else:
            return _deal_with_normal_pay('BC_GATEWAY', '', bank)
    else:
        return _deal_with_normal_pay(channel, '')


def _deal_with_normal_pay(channel, open_id, bank=None):
    # wx js api 需要先获取支付人的open id
    req_params = BCPayReqParams()

    # 当channel为BC_GATEWAY时，bank必填
    if channel == 'BC_GATEWAY':
        req_params.bank = bank

    req_params.channel = channel
    req_params.title = u'python {:s} 支付测试'.format(channel)
    req_params.total_fee = 2
    req_params.bill_no = order_num_on_datetime()
    req_params.optional = {'lang': 'python', u'中文key': u'中文value'}
    # 支付完成后的跳转页面
    req_params.return_url = 'https://beecloud.cn/'
    # 支付宝网页支付(ALI_WEB)的选填参数，可以为商品详细页的url
    # req_params.show_url = 'https://beecloud.cn/'
    # 支付宝内嵌二维码支付(ALI_QRCODE)的必填参数
    if req_params.channel == BCChannelType.ALI_QRCODE:
        req_params.qr_pay_mode = '0'
    if req_params.channel == BCChannelType.WX_JSAPI and open_id:
        req_params.openid = open_id
    resp = bc_pay.pay(req_params)

    if resp.result_code:
        return resp.result_msg + ' # ' + resp.err_detail

    print("beecloud bill object id: " + resp.id)

    if not bc_app.is_test_mode and req_params.channel in [BCChannelType.WX_NATIVE, BCChannelType.BC_NATIVE]:
        return render_template('qrcode.html', raw_content=resp.code_url)
    elif hasattr(resp, 'url') and resp.url:
        print(resp.url)
        return redirect(resp.url)
    elif hasattr(resp, 'html') and resp.html:
        print(resp.html)
        return render_template('blank.html', content=Markup(resp.html))
    elif req_params.channel == BCChannelType.WX_JSAPI:
        jsapi = {}
        jsapi['timeStamp'] = resp.timestamp
        jsapi['appId'] = resp.app_id
        jsapi['nonceStr'] = resp.nonce_str
        jsapi['package'] = resp.package
        jsapi['signType'] = resp.sign_type
        jsapi['paySign'] = resp.pay_sign
        # print(json.dumps(jsapi))
        return render_template('jsapi.html', jsapi=json.dumps(jsapi))


def _deal_with_international_pay(channel):
    req_params = BCInternationalPayParams()
    req_params.channel = channel
    req_params.title = u'python {:s} 支付测试'.format(channel)
    req_params.total_fee = 1
    req_params.currency = 'USD'
    req_params.bill_no = order_num_on_datetime()
    # 支付完成后的跳转页面
    req_params.return_url = 'https://beecloud.cn/'

    resp = bc_pay.international_pay(req_params)

    if resp.result_code:
        return resp.result_msg + ' # ' + resp.err_detail

    if resp.id:
        print("beecloud bill object id: " + resp.id)

    if hasattr(resp, 'url') and resp.url:
        return redirect(resp.url)


@app.route('/bills')
def app_query_bills():
    query_params = BCQueryReqParams()
    if request.args.get('channel', '') != 'ALL':
        query_params.channel = request.args.get('channel', '')
    query_params.spay_result = True

    result = bc_query.query_bills(query_params)
    if result.result_code:
        return result.err_detail

    bills = result.bills
    return render_template('bills.html', bills=bills, channel=request.args.get('channel'))


@app.route('/bills_count')
def app_query_bills_count():
    query_params = BCQueryReqParams()
    if request.args.get('channel', '') != 'ALL':
        query_params.channel = request.args.get('channel', '')
    query_params.spay_result = True

    result = bc_query.query_bills_count(query_params)
    if result.result_code:
        return result.err_detail

    return str(result.count)


@app.route('/refunds')
def app_query_refunds():
    query_params = BCQueryReqParams()
    if request.args.get('channel', '') != 'ALL':
        query_params.channel = request.args.get('channel', '')

    # 过滤退款订单创建时间，此处表示只需要返回北京时间2016-1-13 15:30:00之后的订单
    query_params.start_time = local_timestamp_since_epoch(datetime.datetime(2016, 1, 13, 15, 30, 00))
    query_params.skip = 0
    query_params.limit = 16
    result = bc_query.query_refunds(query_params)
    if result.result_code:
        return result.err_detail
    return render_template('refunds.html', refunds=result.refunds, channel=request.args.get('channel'))


@app.route('/refunds_count')
def app_query_refunds_count():
    query_params = BCQueryReqParams()
    if request.args.get('channel', '') != 'ALL':
        query_params.channel = request.args.get('channel', '')

    # 过滤退款订单创建时间，此处表示只需要返回北京时间2016-1-13 15:30:00之后的订单
    query_params.start_time = local_timestamp_since_epoch(datetime.datetime(2016, 1, 13, 15, 30, 00))
    result = bc_query.query_refunds_count(query_params)
    if result.result_code:
        return result.err_detail

    return str(result.count)


@app.route('/refund_status')
def app_query_refund_status():
    channel = request.args.get('channel')
    refund_no = request.args.get('refund_no')

    result = bc_query.query_refund_status(channel, refund_no)
    if result.result_code:
        return result.err_detail

    return result.refund_status


@app.route('/query_pre_refunds')
def app_query_pre_refunds():
    query_params = BCQueryReqParams()
    if request.args.get('channel', '') != 'ALL':
        query_params.channel = request.args.get('channel', '')
    query_params.need_approval = True
    result = bc_query.query_refunds(query_params)
    if result.result_code:
        return result.err_detail
    return render_template('pre_refunds.html', pre_refunds=result.refunds)


@app.route('/audit_pre_refunds')
def app_audit_pre_refunds():
    req_params = BCPreRefundAuditParams()
    req_params.channel = request.args.get('channel')
    req_params.ids = request.args.get('ids', '').split('@')
    req_params.agree = True
    result = bc_pay.audit_pre_refunds(req_params)
    if not result.result_code:
        # 支付宝退款地址，需用户在支付宝平台上手动输入支付密码处理
        if req_params.channel == BCChannelType.ALI:
            print(result.result_map)
            # url在agree为True时返回
            return redirect(result.url)
        else:
            return str(result.result_map)
    else:
        return result.err_detail


@app.route('/refund')
def app_refund():
    refund_params = BCRefundReqParams()
    # 退款channel为选填参数
    refund_params.channel = request.args.get('channel', '')
    refund_params.refund_no = order_num_on_datetime()
    refund_params.bill_no = request.args.get('bill_no', '')
    refund_params.refund_fee = int(request.args.get('refund_fee', ''))
    if request.args.get('need_approval', ''):
        # 表示预退款，需要后期调用 预退款批量审核 API
        refund_params.need_approval = True
    result = bc_pay.refund(refund_params)
    if not result.result_code:
        # 支付宝退款地址，需用户在支付宝平台上手动输入支付密码处理
        if not refund_params.need_approval and refund_params.channel == BCChannelType.ALI:
            return redirect(result.url)
    return str(result.result_code) + ' # ' + result.result_msg + ' # ' + result.err_detail


@app.route('/bill/id/<bill_id>')
def app_bill_by_id(bill_id):
    result = bc_query.query_bill_by_id(bill_id)
    if result.result_code:
        return result.err_detail
    bill = result.pay
    return render_template('bill.html', bill=bill)


@app.route('/refund/id/<refund_id>')
def app_refund_by_id(refund_id):
    result = bc_query.query_refund_by_id(refund_id)
    if result.result_code:
        return result.err_detail
    m_refund = result.refund
    return render_template('refund_order.html', m_refund=m_refund)


@app.route('/transfer', methods=['POST'])
def app_transfer():
    channel = request.form['channel']
    # 批量打款
    if channel == 'ALI':
        transfer_params = BCBatchTransferParams()
        transfer_params.batch_no = order_num_on_datetime()
        transfer_params.account_name = u'苏州比可网络科技有限公司'
        item1 = BCBatchTransferItem()
        item1.transfer_id = order_num_on_datetime() + 'a'
        item1.receiver_account = '1234567'
        item1.receiver_name = u'某人1'
        item1.transfer_fee = 1
        item1.transfer_note = u'python支付宝批量打款item1'

        item2 = BCBatchTransferItem()
        item2.transfer_id = order_num_on_datetime() + 'b'
        item2.receiver_account = 'ali@account.c'
        item2.receiver_name = 'account2'
        item2.transfer_fee = 1
        item2.transfer_note = u'python支付宝批量打款item2'

        transfer_params.transfer_data = [item1, item2]

        result = bc_pay.batch_transfer(transfer_params)
    elif channel == 'BC_TRANSFER':  # 比可企业打款
        transfer_params = BCCardTransferParams()
        # 单位为分
        transfer_params.total_fee = 1
        transfer_params.bill_no = order_num_on_datetime()
        # 最长支持16个汉字
        transfer_params.title = u'python比可企业打款测试'
        # 银行全名
        transfer_params.bank_fullname = u'中国银行'
        # DE代表借记卡，CR代表信用卡
        transfer_params.card_type = 'DE'
        # 帐户类型，P代表私户，C代表公户
        transfer_params.account_type = 'C'
        # 收款方的银行卡号
        transfer_params.account_no = '5300000'
        # 收款方的姓名或者单位名
        transfer_params.account_name = u'苏州比可网络科技有限公司'
        # 银行绑定的手机号，当需要手机收到银行入账信息时，该值必填，前提是该手机在银行有短信通知业务，否则收不到银行信息
        transfer_params.mobile = '1850000'
        # 附加数据，选填
        transfer_params.optional = {'key1': u'选填的value'}

        result = bc_pay.bc_transfer(transfer_params)

    else:   # 单笔打款
        transfer_params = BCTransferReqParams()
        transfer_params.channel = channel
        transfer_params.desc = u'python {:s} 打款测试'.format(channel)
        if transfer_params.channel == BCChannelType.WX_REDPACK:
            # 微信为10位数字
            transfer_params.transfer_no = order_num_on_datetime()[0:10]
            # 微信红包1.00-200元
            transfer_params.total_fee = 100
            transfer_params.channel_user_id = 'o3kKrjlUsMnv__cK5DYZMl0JoAkY'
            redpack = BCTransferRedPack()
            redpack.send_name = 'BeeCloud'
            redpack.wishing = u'BeeCloud祝福开发者工作顺利'
            redpack.act_name = u'BeeCloud开发者测试中'
            transfer_params.redpack_info = redpack
        elif transfer_params.channel == BCChannelType.WX_TRANSFER:
            # 微信为10位数字
            transfer_params.transfer_no = order_num_on_datetime()[0:10]
            # 微信红包1.00-200元
            transfer_params.total_fee = 100
            transfer_params.channel_user_id = 'o3kKrjlUsMnv__cK5DYZMl0JoAkY'
        else:       # ALI_TRANSFER 支付宝单笔打款
            transfer_params.transfer_no = order_num_on_datetime()
            transfer_params.total_fee = 1
            transfer_params.channel_user_id = 'py@beecloud.cn'
            transfer_params.channel_user_name = 'py'
            transfer_params.account_name = u'苏州比可网络科技有限公司'

        result = bc_pay.transfer(transfer_params)

    if not result.result_code:
        print("beecloud transfer object id: " + result.id)
    
    if hasattr(result, 'url') and result.url:
        # 支付宝需要跳转到如下支付宝链接输入支付密码确认
        return redirect(result.url)

    return str(result.result_code) + ' # ' + result.result_msg + ' # ' + result.err_detail


@app.route('/subscribe', methods=['POST'])
def subscribe():
    param = BCSubscription()
    for k, v in request.form.to_dict().items():
        if not k.startswith('sms'):
            setattr(param, k, v)

    result = bc_subscribe.subscribe(param, request.form['sms_id'], request.form['sms_code'])
    if result.result_code:
        return u'订阅失败：' + result.result_msg + " | " + result.err_detail
    else:
        return u'订阅请求成功，请注意webhook接收最终审核结果：' + result.subscription.id


@app.route('/subscription/banks')
def subscription_supported_banks():
    result = bc_query.query_subscription_payment_supported_banks()
    if not result.result_code:
        return json.dumps(result.common_banks)
    else:
        return '[]'


@app.route('/subscription/plans')
def subscription_plans():
    # 自定义你的查询条件
    param = BCQueryCriteria()
    param.limit = 15
    result = bc_query.query_plans(param)
    if not result.result_code:
        data = [{'id': plan.id, 'name': plan.name} for plan in result.plans if plan.valid]
        return json.dumps(data)
    else:
        return '{}'


@app.route('/sms')
def sms():
    mobile = request.args.get('mobile')
    print(mobile)
    result = send_sms_passcode(bc_app, mobile)
    if not result.result_code:
        print('sms id:' + result.sms_id)
        return result.sms_id
    else:
        print(result.result_msg)
        print(result.err_detail)
        return ''


@app.route('/subscriptions')
def subscriptions():
    # 自定义你的查询条件
    param = BCQueryCriteria()
    param.buyer_id = 'xz1'
    # 作为query_subscriptions的参数

    result = bc_query.query_subscriptions(param)
    return render_template('subscriptions.html', subscriptions=result.subscriptions)


@app.route('/cancel_subscription')
def cancel_subscription():
    sid = request.args.get('sid')
    if not sid:
        return ''

    result = bc_subscribe.cancel_subscription(sid)
    if result.result_code:
        return 'cancel failed, reason: ' + result.result_msg + ' | ' + result.err_detail
    else:
        return 'cancel succ, subscription id: ' + result.id


@app.template_filter('format_utc_time')
def format_utc_time(s):
    return datetime.datetime.fromtimestamp(s/1000.0).strftime('%Y-%m-%d %H:%M:%S')

if __name__ == '__main__':
    app.debug = True
    # app.run(host='pythondemo.beecloud.cn', port=80)
    app.run()
